essentials-framework/Essentials Core/PepperDashEssentialsBase/Factory/DeviceFactory.cs

```diff
@@ -83,33 +85,78 @@ public static void AddFactoryForType(string typeName, string description, CType
-            if (FactoryMethods.ContainsKey(typeName))
-            {
-                Debug.Console(0, Debug.ErrorLogLevel.Notice, "Loading '{0}' from Essentials Core", dc.Type);
-                return FactoryMethods[typeName].FactoryMethod(dc);
-            }
-
-            return null;
-        }
-
+        }
+
+        private static void CheckForSecrets(IEnumerable<JProperty> obj)
+        {
+            foreach (var prop in obj.Where(prop => prop.Value as JObject != null))
+            {
+                if (prop.Name.ToLower() == "secret")
+                {
+                    var secret = GetSecret(JsonConvert.DeserializeObject<SecretsPropertiesConfig>(prop.Children().First().ToString()));
+                    prop.Parent.Replace(secret);
+                }
+                var recurseProp = prop.Value as JObject;
+                if (recurseProp == null) return;
+                CheckForSecrets(recurseProp.Properties());
+            }
+        }
+
+        private static string GetSecret(SecretsPropertiesConfig data)
+        {
+            var secretProvider = SecretsManager.GetSecretProviderByKey(data.Provider);
+            if (secretProvider == null) return null;
+            var secret = secretProvider.GetSecret(data.Key);
+            if (secret != null) return (string) secret.Value;
+            Debug.Console(1,
+                "Unable to retrieve secret {0}{1} - Make sure you've added it to the secrets provider",
+                data.Provider, data.Key);
+            return String.Empty;
+        }
+
+
+        /// <summary>
+        /// The factory method for Core "things". Also iterates the Factory methods that have
+        /// been loaded from plugins
+        /// </summary>
+        /// <param name="dc"></param>
+        /// <returns></returns>
+        public static IKeyed GetDevice(DeviceConfig dc)
+        {
+            try
+            {
+                Debug.Console(0, Debug.ErrorLogLevel.Notice, "Loading '{0}' from Essentials Core", dc.Type);
+
+                var localDc = new DeviceConfig(dc);
+
+                var key = localDc.Key;
+                var name = localDc.Name;
+                var type = localDc.Type;
+                var properties = localDc.Properties;
+                //var propRecurse = properties;
+
+                var typeName = localDc.Type.ToLower();
+
+
+                var jObject = properties as JObject;
+                if (jObject != null)
+                {
+                    var jProp = jObject.Properties();
+
+                    CheckForSecrets(jProp);
+                }
+
+                Debug.Console(2, "typeName = {0}", typeName);
+                // Check for types that have been added by plugin dlls. 
+                return !FactoryMethods.ContainsKey(typeName) ? null : FactoryMethods[typeName].FactoryMethod(localDc);
+            }
+            catch (Exception ex)
+            {
+                Debug.Console(2, "Issue with getting device - {0}", ex.Message);

```

Comment:
Use this pattern with these sorts of Exceptions that might occur during startup:
```
Debug.Console(0, Debug.ErrorLogLevel.Error, "Exception occurred while creating device {0}: {1}", dc.Key, ex.message);
Debug.Console(2, "{0}", ex.StackTrace);
```
This allows for easier debugging when something fails without having to go all the way to appdebug 2 to even see the error message.